File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)

    def dfs(node, a, b, visited):
        visited[node] = True
        count = 1
        
        for nxt in graph[node]:
            if (node == a and nxt == b) or (node == b and nxt == a):
                continue
            
            if not visited[nxt]:
                count += dfs(nxt, a, b, visited)
        return count

    answer = float('inf')

    for a, b in wires:
        visited = [False] * (n+1)
        count = dfs(a, a, b, visited)
        diff = abs((n - count) - count)
        answer = min(answer, diff)

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    static List<Integer>[] graph;
    static boolean[] visited;

    public int solution(int n, int[][] wires) {
        graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        for (int[] wire : wires) {
            int a = wire[0];
            int b = wire[1];
            graph[a].add(b);
            graph[b].add(a);
        }

        int answer = Integer.MAX_VALUE;

        for (int[] wire : wires) {
            int a = wire[0];
            int b = wire[1];

            visited = new boolean[n + 1];

            int count = dfs(a, a, b); 

            int diff = Math.abs((n - count) - count);
            answer = Math.min(answer, diff);
        }

        return answer;
    }

    private int dfs(int node, int cutA, int cutB) {
        visited[node] = true;
        int count = 1;

        for (int nxt : graph[node]) {
            if ((node == cutA && nxt == cutB) || (node == cutB && nxt == cutA)) {
                continue;
            }

            if (!visited[nxt]) {
                count += dfs(nxt, cutA, cutB);
            }
        }

        return count;
    }
}